tools/codegen-crds/yaml_patch.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+import (
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	kyaml "sigs.k8s.io/yaml"
+
+	yamlpatch "github.com/krishicks/yaml-patch"
```

Comment:
I'd prefer to integrate at the level of binaries because I think there is less risk of dependency conflict and I think it's more likely that we stay up to date.  Is this allowing us to do something we cannot otherwise achieve?

Reply:
> I think there is less risk of dependency conflict and I think it's more likely that we stay up to date. Is this allowing us to do something we cannot otherwise achieve?

I have a few thoughts here:
* We are already importing a lot of the dependencies that we expect to go into this project going forward, all of `k8s.io/codegen` for example, so with those, the dependency question is somewhat null
* If we fork/exec, it gives us less control over the errors and the error handling IMO, here we can clearly see what we are doing, where, and which stage went wrong, the command we are replacing was `yaml-patch -o patch.yaml < original.file > output.file && mv output.file > original.file`, with all the indirection there it's hard to know exactly what went wrong
* With the schemapatch and deepycopy/etc generators, an advantage of having them in-process like this is that when we parse the package roots (which I've seen to be very slow on occasion) we can share the parsed package roots between the generators. We had an issue recently where the roots were taking ages to parse causing the generate to be slow, the speed advantage here should make the process better for everyone
* I worry about our ability to stay up to date if these *aren't* in tree. With the current approach we know they're compatible because the compiler would barf if there are ever any changes that we need to resolve. Recently YQ changed from v2 to v3, if we were to allow users to specify their own YQ binary, we don't know which version of that they are using, and they have completely different CLIs. If we ever want to make updates, I think it's easier to manage that centrally and provide a single binary to our consumers/api developers rather than having to make sure they clean their bin dir every time
* If we interface at the binary layer, we still need to make sure all our users have the right binaries installed and then somehow keep those up to date, which seems harder than risking the dependency issues here
* For dependency issues, the only issue I can see is the yaml-patch library which is currently unmaintained, we are maintaining controller-tools so that can/will stay up to date, and if we see issues with the yaml-patch library, well it's pretty small, perhaps we should fork that too?

Reply:
I'd rather not grow forks unless I have to.  One thing to consider is that usage documentation is usually written at the level of binaries.  Embedding golang code and wiring in a library can make it more difficult for unfamiliar users to figure out what's wrong.  Something to consider and balance.

Reply:
> I'd rather not grow forks unless I have to

AFAIK the only thing we are forking is the controller-tools project, everything else is coming from the original upstream apart from perhaps the yaml-patch as that library is archived (we should probably find a maintained version)

> One thing to consider is that usage documentation is usually written at the level of binaries. Embedding golang code and wiring in a library can make it more difficult for unfamiliar users to figure out what's wrong.

In my experience recently, the help text for these tools is next to useless. By writing a wrapper around this, we can, as the "experts" in the tooling determine ahead of time the appropriate inputs and expected values and verify this input data ahead of time, providing useful feedback to our end users.

For example, even within this PR, the code checks for valid YAML unmarshalling and reports back errors if the YAML isn't structured correctly. The schemapatch tool doesn't do this and ignores anything that doesn't unmarshal correctly, you don't want to know how much time I lost because of that.

I appreciate the docs are written typically for the binary interaction, but we are building an abstraction on top to make it easier for users, I see a lot of value in pre-defining validations before we pass the input to the generation library